### PR TITLE
Terraria: Add crate logic

### PR DESCRIPTION
**Pull request by Seldom-SE**
## What is this fixing or adding?

Terraria fishing crate logic. I originally excluded it because it could be grindy, but I've found that it doesn't affect accessibility anyway, except for the Head in the Clouds achievement, whose logic I've changed based on whether grindy achievements are enabled.

## How was this tested?

I generated a world and made sure checks still work, and then I generated a lot of worlds and checked if their spoilers match what I expect

### CODE REVIEW

**Review by benny-dreamly — APPROVED** [review]

File: worlds/terraria/Checks.py [hunk 1]
```diff
@@ -221,62 +222,60 @@ def mark_progression(
...
-        List[str],
-        # Progression rules
-        Set[str],
-        # Armor to minion count,
-        Dict[str, int],
-        # Accessory to minion count,
-        Dict[str, int],
-    ]
-):
+def read_data() -> Tuple[
```
Comment (agilbert1412) [inline worlds/terraria/Checks.py:225]: Does this monster have to be built out of recursive tuples?
I'm scared

Comment (benny-dreamly) [reply worlds/terraria/Checks.py:225]: Oh god I missed that now I’m shaking in my boots

Comment (benny-dreamly) [reply worlds/terraria/Checks.py:225]: there's a reason for it. it accurately represents the data structure of the output data or something idk that's chatgpts word

Comment (Seldom-SE) [reply worlds/terraria/Checks.py:225]: I think the outer tuple is fine, it gets split up on line 719, but yeah the rules type is messy. It's not relevant to this pr, though. It just got caught by my formatter. I'll move it to a class if I end up doing parser reworks

**Review by Exempt-Medic — APPROVED** [review]
Changes LGTM. Tested the grindy logic by looking at spoiler logs. Walked through the logic changes with a friend who knows the game better than myself; we concluded that they were correct.

**Review by Berserker66 — APPROVED** [review]

**Pull request merged**